src/libsystemd/sd-bus/bus-message.c

```diff
@@ -889,7 +890,78 @@ int bus_message_new_synthetic_error(
         return 0;
 }
 
-DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC(sd_bus_message, sd_bus_message, message_free);
+
+_public_ sd_bus_message* sd_bus_message_ref(sd_bus_message *m) {
+        if (!m)
+                return NULL;
+
+        /* We are fine if this message so far was either explicitly reffed or not reffed but queued into at
+         * least one bus connection object. */
+        assert(m->n_ref > 0 || m->n_queued > 0);
+
+        m->n_ref++;
+
+        /* Each user reference to a bus message shall also be considered a ref on the bus */
+        sd_bus_ref(m->bus);
+        return m;
+}
+
+_public_ sd_bus_message* sd_bus_message_unref(sd_bus_message *m) {
+        if (!m)
+                return NULL;
+
+        assert(m->n_ref > 0);
+
+        sd_bus_unref(m->bus); /* Each regular ref is also a ref on the bus connection. Let's hence drop it
+                               * here. Note we have to do this before decrementing our own n_ref here, since
+                               * otherwise, if this message is currently queued sd_bus_unref() might call
+                               * bus_message_unref_queued() for this which might then destroy the message
+                               * while we are still processing it. */
+        m->n_ref--;
```

Comment:
the explanation for this is supposed to be in the comment a bit further down: we only reset m->bus when the last ref is dropped. 

The idea here is that every ref onto a message als refs the bus implicitly. But this also means that each unref should unref the bus implicitly, which is what you see here. So, if you have two refs on the msg, then this means two refs on the bus from the msg and if you then unref the first of that one the bus field needs to stay around so that the second unref still can unref it. Only after all our own refs are gone the bus object becomes invalidated to us, and only then we reset the field a bit further down.

Reply:
will add a brief comment about that.